src/eth1spec/rlp.py

```diff
@@ -1,274 +1,328 @@
 from .eth_types import Bytes
 
-verbose = False
-debug = False
-
-
 RLP = Union[Bytes, Uint, U256, Sequence["RLP"]]  # type: ignore
 
 
-def encode(x: RLP) -> Bytes:
-    """
-    Encodes `x` into a sequence of bytes using RLP.
-
-    Parameters
-    ----------
-    x : `RLP`
-        A `Bytes`, `Uint`, or sequence of `RLP` encodable objects.
-
-    Returns
-    -------
-    encoded : `eth1spec.eth_types.Bytes`
-        The RLP encoded bytes representing `x`.
-    """
-    if verbose:
-        print("RLP(", x, ")", "type: ", type(x))
-    if isinstance(x, (bytearray, bytes)):
-        return R_b(x)
-    elif isinstance(x, (Uint, U256)):
-        return encode(x.to_be_bytes())
-    elif isinstance(x, Sequence):
-        return R_l(cast(Sequence[RLP], x))
-    else:
-        raise TypeError()
-
-
-# binary encoding/decoding
-def R_b(x: Bytes) -> Bytes:
-    """
-    Encodes `x`, a sequence of bytes, using RLP.
-
-    Parameters
-    ----------
-    x : `eth1spec.eth_types.Bytes`
-        Bytes to encode with RLP.
-
-    Returns
-    -------
-    encoded : `eth1spec.eth_types.Bytes`
-        The RLP encoded bytes representing `x`.
-    """
-    if verbose:
-        print("R_b(", x, ")")
-    len_x = Uint(len(x))
-    if len_x == 1 and x[0] < 128:
-        return x  # bytearray([x[0] + 0x80]) # noqa: SC100
-    elif len_x < 56:
-        return bytearray([128 + len_x]) + x
-    else:
-        big_endian = len_x.to_be_bytes()
-        return bytearray([183 + len(big_endian)]) + big_endian + x
-
-
-# list encoding/decoding
-def R_l(x: Sequence[RLP]) -> Bytes:
-    """
-    Encodes `x`, a list of RLP encodable objects, using RLP.
-
-    Parameters
-    ----------
-    x : `List[RLP]`
-        List to encode with RLP.
-
-    Returns
-    -------
-    encoded : `eth1spec.eth_types.Bytes`
-        The RLP encoded bytes representing `x`.
-    """
-    if verbose:
-        print("R_l(", x, ")")
-    partial_x = s(x)
-    len_partial_x = Uint(len(partial_x))
-    if len_partial_x < 56:
-        return bytearray([192 + len_partial_x]) + partial_x
-    else:
-        big_endian = len_partial_x.to_be_bytes()
-        return bytearray([247 + len(big_endian)]) + big_endian + partial_x
-
-
-# for a list, recursively call RLP or RLP_inverse
-def s(x: Sequence[RLP]) -> Bytes:
-    """
-    Partially encodes `x`, a list of RLP encodable objects, using RLP,
-    excluding the length prefix.
-
-    Parameters
-    ----------
-    x : `List[RLP]`
-        List to encode with RLP.
-
-    Returns
-    -------
-    partially_encoded : `eth1spec.eth_types.Bytes`
-        The partially RLP encoded bytes representing `x`.
-    """
-    if verbose:
-        print("s(", x, ")")
-    partially_encoded = bytearray([])
-    for xi in x:
-        partially_encoded += encode(xi)
-    return partially_encoded
-
-
-# inverses of above
-
-
-def decode(b: Bytes) -> RLP:
+# TODO:
+#  Write testcases
+#  Run the testcases against fixtures
+class RLPEncoder(object):
```

Comment:
This is kind of a data class. It has only classmethods, and the only reason I am using a class here, is to kind of group the functionalities into a specific purpose like encoding/decoding. Let me know if this will disrupt the readability. Thanks.

Reply:
Besides the types in `base_types.py` which need to be object oriented to override the arithmetic operators, classes shouldn't have _any_ functions/methods.

_discussion continued in call_

Reply:
Removed the class based structure with normal functions.